examples/vision/semisupervised_simclr.py

```diff
@@ -0,0 +1,743 @@
+and does not have to be reimplemented when deploying it
+"""
+
+# the implementation of these image augmentations follow the torchvision library:
+# https://github.com/pytorch/vision/blob/master/torchvision/transforms/transforms.py
+# https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional_tensor.py
+
+# however these augmentations:
+# -run on batches of images
+# -run on gpu
+# -can be part of a model
+
+
+# crops and resizes part of the image to the original resolutions
+class RandomResizedCrop(layers.Layer):
+    def __init__(self, scale, ratio, **kwargs):
+        super().__init__(**kwargs)
+        # area-range of the cropped part: (min area, max area), uniform sampling
+        self.scale = scale
+        # aspect-ratio-range of the cropped part: (log min ratio, log max ratio), log-uniform sampling
+        self.log_ratio = (tf.math.log(ratio[0]), tf.math.log(ratio[1]))
+
+    def call(self, images, training=True):
+        if training:
+            batch_size = tf.shape(images)[0]
+            height = tf.shape(images)[1]
+            width = tf.shape(images)[2]
+
+            # independently sampled scales and ratios for every image in the batch
+            random_scales = tf.random.uniform(
+                (batch_size,), self.scale[0], self.scale[1]
+            )
+            random_ratios = tf.exp(
+                tf.random.uniform((batch_size,), self.log_ratio[0], self.log_ratio[1])
+            )
+
+            # corresponding height and widths, clipped to fit in the image
+            new_heights = tf.clip_by_value(tf.sqrt(random_scales / random_ratios), 0, 1)
+            new_widths = tf.clip_by_value(tf.sqrt(random_scales * random_ratios), 0, 1)
+
+            # random anchors for the crop bounding boxes
+            height_offsets = tf.random.uniform((batch_size,), 0, 1 - new_heights)
+            width_offsets = tf.random.uniform((batch_size,), 0, 1 - new_widths)
+
+            # assemble bounding boxes and crop
+            bounding_boxes = tf.stack(
+                [
+                    height_offsets,
+                    width_offsets,
+                    height_offsets + new_heights,
+                    width_offsets + new_widths,
+                ],
+                axis=1,
+            )
+            images = tf.image.crop_and_resize(
+                images, bounding_boxes, tf.range(batch_size), (height, width)
+            )
+
+        return images
+
+
+# distorts the color distibutions of images
+class RandomColorJitter(layers.Layer):
+    def __init__(self, brightness=0, contrast=0, saturation=0, hue=0, **kwargs):
+        super().__init__(**kwargs)
+
+        # color jitter ranges: (min jitter strength, max jitter strength)
+        self.brightness = brightness
+        self.contrast = contrast
+        self.saturation = saturation
+        self.hue = hue
+
+        # list of applicable color augmentations
+        self.color_augmentations = [
```

Comment:
All this really does is apply a random affine transform. You can implement a random affine transform in ~15 lines top, which will save you quite a few lines.